### **Preprocessing of Flats Dataset**

In [16]:
# importing the required libraries
import pandas as pd
import numpy as np
import re

In [7]:
# importing the flats dataset
data = pd.read_csv('../data/flats.csv')

In [ ]:
# checking first 5 rows of the dataset
data.head()

property_name  \
0     2 BHK Flat in Krishna Colony   
1        2 BHK Flat in Ashok Vihar   
2              2 BHK Flat in Sohna   
3  2 BHK Flat in Sector 61 Gurgaon   
4  2 BHK Flat in Sector 92 Gurgaon   

                                                link                 society  \
0  https://www.99acres.com/2-bhk-bedroom-apartmen...  maa bhagwati residency   
1  https://www.99acres.com/2-bhk-bedroom-apartmen...            Apna Enclave   
2  https://www.99acres.com/2-bhk-bedroom-apartmen...  Tulsiani Easy in Homes   
3  https://www.99acres.com/2-bhk-bedroom-apartmen...     Smart World Orchard   
4  https://www.99acres.com/2-bhk-bedroom-apartmen...        Parkwood Westend   

        price             area                            areaWithType  \
0      45 Lac   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)   
1      50 Lac   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)   
2      40 Lac   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)   
3  1.47 Crore  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)   
4      70 Lac   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)   

      bedRoom     bathroom      balcony additionalRoom  \
0  2 Bedrooms  2 Bathrooms    1 Balcony            NaN   
1  2 Bedrooms  2 Bathrooms    1 Balcony            NaN   
2  2 Bedrooms  2 Bathrooms  3 Balconies            NaN   
3  2 Bedrooms  2 Bathrooms  2 Balconies     Study Room   
4  2 Bedrooms  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession                                    nearbyLocations  \
0     1 to 5 Year Old  ['Chintapurni Mandir', 'State bank ATM', 'Pear...   
1        10+ Year Old  ['Chintapurni Mandir', 'Sheetla Mata Mandir', ...   
2     0 to 1 Year Old  ['Huda City Metro', 'Golf Course extn road', '...   
3              Dec-23  ['Sector 55-56 Metro station', 'Bestech Centra...   
4  Under Construction  ['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...   

                                         description  \
0  So with lift.Maa bhagwati residency is one of ...   
1  Property situated on main road, railway statio...   
2  This property is 15 km away from badshapur, gu...   
3  Near to metro station of sector 56 and opposit...   
4  We are the proud owners of this 2 bhk alongwit...   

                                      furnishDetails  \
0  ['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...   
1  ['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...   
2                                                NaN   
3                                                NaN   
4                                                 []   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Security / F...   
1  ['Security / Fire Alarm', 'Maintenance Staff',...   
2  ['Power Back-up', 'Feng Shui / Vaastu Complian...   
3  ['Security / Fire Alarm', 'Private Garden / Te...   
4                                                NaN   

                                              rating property_id  
0  ['Environment4 out of 5', 'Safety4 out of 5', ...   C68850746  
1  ['Environment4 out of 5', 'Safety4 out of 5', ...   H68850564  
2  ['Environment4 out of 5', 'Safety4 out of 5', ...   J68850120  
3                                                NaN   S68849476  
4  ['Environment5 out of 5', 'Safety3 out of 5', ...   L47956793

In [10]:
# checking for the missing values
data.isnull().sum()

property_name         0
link                  0
society               1
price                 1
area                 12
areaWithType          0
bedRoom               0
bathroom              0
balcony               0
additionalRoom     1314
address               6
floorNum              2
facing              881
agePossession         1
nearbyLocations      95
description           0
furnishDetails      805
features            414
rating              332
property_id           0
dtype: int64

In [9]:
# checking the data info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3009 entries, 0 to 3008
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    3009 non-null   object
 1   link             3009 non-null   object
 2   society          3008 non-null   object
 3   price            3008 non-null   object
 4   area             2997 non-null   object
 5   areaWithType     3009 non-null   object
 6   bedRoom          3009 non-null   object
 7   bathroom         3009 non-null   object
 8   balcony          3009 non-null   object
 9   additionalRoom   1695 non-null   object
 10  address          3003 non-null   object
 11  floorNum         3007 non-null   object
 12  facing           2128 non-null   object
 13  agePossession    3008 non-null   object
 14  nearbyLocations  2914 non-null   object
 15  description      3009 non-null   object
 16  furnishDetails   2204 non-null   object
 17  features         2595 non-null   

In [11]:
# checking for duplicates in the data
data.duplicated().sum()

np.int64(0)

There are no duplicate values in the data.

There is no need of the `link` and `property_id` feature. Thus removing these features.

In [12]:
# droping link and property_id
data.drop(columns=['link','property_id'], inplace=True)

In [13]:
data.head()

property_name                 society       price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency      45 Lac   
1        2 BHK Flat in Ashok Vihar            Apna Enclave      50 Lac   
2              2 BHK Flat in Sohna  Tulsiani Easy in Homes      40 Lac   
3  2 BHK Flat in Sector 61 Gurgaon     Smart World Orchard  1.47 Crore   
4  2 BHK Flat in Sector 92 Gurgaon        Parkwood Westend      70 Lac   

              area                            areaWithType     bedRoom  \
0   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession                                    nearbyLocations  \
0     1 to 5 Year Old  ['Chintapurni Mandir', 'State bank ATM', 'Pear...   
1        10+ Year Old  ['Chintapurni Mandir', 'Sheetla Mata Mandir', ...   
2     0 to 1 Year Old  ['Huda City Metro', 'Golf Course extn road', '...   
3              Dec-23  ['Sector 55-56 Metro station', 'Bestech Centra...   
4  Under Construction  ['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...   

                                         description  \
0  So with lift.Maa bhagwati residency is one of ...   
1  Property situated on main road, railway statio...   
2  This property is 15 km away from badshapur, gu...   
3  Near to metro station of sector 56 and opposit...   
4  We are the proud owners of this 2 bhk alongwit...   

                                      furnishDetails  \
0  ['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...   
1  ['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...   
2                                                NaN   
3                                                NaN   
4                                                 []   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Security / F...   
1  ['Security / Fire Alarm', 'Maintenance Staff',...   
2  ['Power Back-up', 'Feng Shui / Vaastu Complian...   
3  ['Security / Fire Alarm', 'Private Garden / Te...   
4                                                NaN   

                                              rating  
0  ['Environment4 out of 5', 'Safety4 out of 5', ...  
1  ['Environment4 out of 5', 'Safety4 out of 5', ...  
2  ['Environment4 out of 5', 'Safety4 out of 5', ...  
3                                                NaN  
4  ['Environment5 out of 5', 'Safety3 out of 5', ...

In [14]:
# checking the shape of the dataset
data.shape

(3009, 18)

After removing 2 features, current dataset has 3009 rows and 18 columns.

## **Preprocessing**

#### **1. `society` Column**

In [15]:
data['society'].value_counts()

society
SS The Leaf3.8 ★                             73
Tulip Violet4.3 ★                            40
Shapoorji Pallonji Joyville Gurugram4.0 ★    39
Signature Global Park4.0 ★                   36
Shree Vardhman Victoria3.8 ★                 35
                                             ..
Chanderlok Society                            1
Ansal Estella                                 1
Laxmi Apartment sector 99A Gurgaon            1
smart world one                               1
Sanskriti Apartment                           1
Name: count, Length: 637, dtype: int64

There are many values that ends with a `★`, which is not needed.

In [18]:
data['society'] = data['society'].apply(lambda name: re.sub(r'\d+(\.\d+)?\s?★', '', str(name)).strip()).str.lower()

In [19]:
data['society'].value_counts()

society
tulip violet                            75
ss the leaf                             74
shapoorji pallonji joyville gurugram    45
dlf new town heights                    38
signature global park                   36
                                        ..
bhemerlei                                1
wings apartment                          1
uphaar residency 2                       1
the center court                         1
divya apartments                         1
Name: count, Length: 603, dtype: int64

In [21]:
data.head()

property_name                 society       price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency      45 Lac   
1        2 BHK Flat in Ashok Vihar            apna enclave      50 Lac   
2              2 BHK Flat in Sohna  tulsiani easy in homes      40 Lac   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard  1.47 Crore   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend      70 Lac   

              area                            areaWithType     bedRoom  \
0   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession                                    nearbyLocations  \
0     1 to 5 Year Old  ['Chintapurni Mandir', 'State bank ATM', 'Pear...   
1        10+ Year Old  ['Chintapurni Mandir', 'Sheetla Mata Mandir', ...   
2     0 to 1 Year Old  ['Huda City Metro', 'Golf Course extn road', '...   
3              Dec-23  ['Sector 55-56 Metro station', 'Bestech Centra...   
4  Under Construction  ['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...   

                                         description  \
0  So with lift.Maa bhagwati residency is one of ...   
1  Property situated on main road, railway statio...   
2  This property is 15 km away from badshapur, gu...   
3  Near to metro station of sector 56 and opposit...   
4  We are the proud owners of this 2 bhk alongwit...   

                                      furnishDetails  \
0  ['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...   
1  ['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...   
2                                                NaN   
3                                                NaN   
4                                                 []   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Security / F...   
1  ['Security / Fire Alarm', 'Maintenance Staff',...   
2  ['Power Back-up', 'Feng Shui / Vaastu Complian...   
3  ['Security / Fire Alarm', 'Private Garden / Te...   
4                                                NaN   

                                              rating  
0  ['Environment4 out of 5', 'Safety4 out of 5', ...  
1  ['Environment4 out of 5', 'Safety4 out of 5', ...  
2  ['Environment4 out of 5', 'Safety4 out of 5', ...  
3                                                NaN  
4  ['Environment5 out of 5', 'Safety3 out of 5', ...

#### **2. `price` Column**

In [22]:
data['price'].value_counts()

price
1.25 Crore    79
1.1 Crore     61
1.4 Crore     60
1.2 Crore     59
1.5 Crore     59
              ..
26.9 Lac       1
26.5 Lac       1
73.53 Lac      1
89.77 Lac      1
62.5 Lac       1
Name: count, Length: 440, dtype: int64

In `price` column some flats have there price in Lacs and some have there price in Crore. Thus converting the flats price in a single unit that is Crore. 

Removing flats that have values as `Price on Request`. 

In [24]:
data = data[data['price']!='Price on Request']

In [30]:
# Creating a price convertor function
def price_converter(x):
    """
    This function converts price amounts into Crores.
    """
    # If the input is already a float, return it directly
    if isinstance(x, float):
        return x

    # Split the string into parts
    parts = x.split(' ')

    # Check if the split resulted in at least two parts
    if len(parts) < 2:
        return None  # Or handle it in another way, like returning 0

    # Extract the numeric value and unit
    numeric_value = float(parts[0])
    unit = parts[1]

    # Convert based on the unit
    if unit == 'Lac':
        return round(numeric_value / 100, 2)
    elif unit == 'Crore':
        return round(numeric_value, 2)
    else:
        return None  # Handle unexpected units

In [32]:
# Applying the price converter function to the price feature
data['price'] = data['price'].apply(price_converter)

C:\Users\aasth\AppData\Local\Temp\ipykernel_18516\3077879571.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['price'] = data['price'].apply(price_converter)


In [33]:
data.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

              area                            areaWithType     bedRoom  \
0   ₹ 5,000/sq.ft.          Carpet area: 900 (83.61 sq.m.)  2 Bedrooms   
1   ₹ 7,692/sq.ft.          Carpet area: 650 (60.39 sq.m.)  2 Bedrooms   
2   ₹ 6,722/sq.ft.          Carpet area: 595 (55.28 sq.m.)  2 Bedrooms   
3  ₹ 12,250/sq.ft.        Carpet area: 1200 (111.48 sq.m.)  2 Bedrooms   
4   ₹ 5,204/sq.ft.  Super Built up area 1345(124.95 sq.m.)  2 Bedrooms   

      bathroom      balcony additionalRoom  \
0  2 Bathrooms    1 Balcony            NaN   
1  2 Bathrooms    1 Balcony            NaN   
2  2 Bathrooms  3 Balconies            NaN   
3  2 Bathrooms  2 Balconies     Study Room   
4  2 Bathrooms  3 Balconies     Study Room   

                               address             floorNum facing  \
0     Krishna Colony, Gurgaon, Haryana    4th   of 4 Floors   West   
1   46b, Ashok Vihar, Gurgaon, Haryana    1st   of 3 Floors   West   
2              Sohna, Gurgaon, Haryana  12nd   of 14 Floors    NaN   
3  Sector 61 Gurgaon, Gurgaon, Haryana    2nd   of 4 Floors    NaN   
4  Sector 92 Gurgaon, Gurgaon, Haryana    5th   of 8 Floors    NaN   

        agePossession                                    nearbyLocations  \
0     1 to 5 Year Old  ['Chintapurni Mandir', 'State bank ATM', 'Pear...   
1        10+ Year Old  ['Chintapurni Mandir', 'Sheetla Mata Mandir', ...   
2     0 to 1 Year Old  ['Huda City Metro', 'Golf Course extn road', '...   
3              Dec-23  ['Sector 55-56 Metro station', 'Bestech Centra...   
4  Under Construction  ['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...   

                                         description  \
0  So with lift.Maa bhagwati residency is one of ...   
1  Property situated on main road, railway statio...   
2  This property is 15 km away from badshapur, gu...   
3  Near to metro station of sector 56 and opposit...   
4  We are the proud owners of this 2 bhk alongwit...   

                                      furnishDetails  \
0  ['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...   
1  ['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...   
2                                                NaN   
3                                                NaN   
4                                                 []   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Security / F...   
1  ['Security / Fire Alarm', 'Maintenance Staff',...   
2  ['Power Back-up', 'Feng Shui / Vaastu Complian...   
3  ['Security / Fire Alarm', 'Private Garden / Te...   
4                                                NaN   

                                              rating  
0  ['Environment4 out of 5', 'Safety4 out of 5', ...  
1  ['Environment4 out of 5', 'Safety4 out of 5', ...  
2  ['Environment4 out of 5', 'Safety4 out of 5', ...  
3                                                NaN  
4  ['Environment5 out of 5', 'Safety3 out of 5', ...

#### **3. `area` Column**

The `area` column is having information regaarding price per square feet thus changing the name of the column to `price_per_sqft`

In [34]:
data.rename(columns={'area':'price_per_sqft'}, inplace=True)

In [35]:
data.head(2)

property_name                 society  price  \
0  2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1     2 BHK Flat in Ashok Vihar            apna enclave   0.50   

   price_per_sqft                    areaWithType     bedRoom     bathroom  \
0  ₹ 5,000/sq.ft.  Carpet area: 900 (83.61 sq.m.)  2 Bedrooms  2 Bathrooms   
1  ₹ 7,692/sq.ft.  Carpet area: 650 (60.39 sq.m.)  2 Bedrooms  2 Bathrooms   

     balcony additionalRoom                             address  \
0  1 Balcony            NaN    Krishna Colony, Gurgaon, Haryana   
1  1 Balcony            NaN  46b, Ashok Vihar, Gurgaon, Haryana   

            floorNum facing    agePossession  \
0  4th   of 4 Floors   West  1 to 5 Year Old   
1  1st   of 3 Floors   West     10+ Year Old   

                                     nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pear...   
1  ['Chintapurni Mandir', 'Sheetla Mata Mandir', ...   

                                         description  \
0  So with lift.Maa bhagwati residency is one of ...   
1  Property situated on main road, railway statio...   

                                      furnishDetails  \
0  ['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...   
1  ['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Security / F...   
1  ['Security / Fire Alarm', 'Maintenance Staff',...   

                                              rating  
0  ['Environment4 out of 5', 'Safety4 out of 5', ...  
1  ['Environment4 out of 5', 'Safety4 out of 5', ...

This column have **₹** in the starting of the values and **/sqft** at the end. Removing these both.

In [47]:
data[data['price_per_sqft']=='area']

property_name  society  price price_per_sqft  areaWithType  bedRoom  \
286  property_name  society    NaN           area  areaWithType  bedRoom   

     bathroom  balcony  additionalRoom  address  floorNum  facing  \
286  bathroom  balcony  additionalRoom  address  floorNum  facing   

     agePossession  nearbyLocations  description  furnishDetails  features  \
286  agePossession  nearbyLocations  description  furnishDetails  features   

     rating  
286  rating

In [48]:
data.drop(index=286,inplace=True)

In [51]:
data['price_per_sqft'] = data['price_per_sqft'].str.split('/').str.get(0).str.replace('₹','').str.replace(',','').str.strip().astype('float')

In [52]:
data['price_per_sqft'].value_counts()

price_per_sqft
10000.0    19
12500.0    16
8000.0     16
6666.0     13
5000.0     13
           ..
4723.0      1
6129.0      1
7234.0      1
8823.0      1
16859.0     1
Name: count, Length: 2128, dtype: int64

In [53]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2997 entries, 0 to 3008
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    2997 non-null   object 
 1   society          2997 non-null   object 
 2   price            2996 non-null   float64
 3   price_per_sqft   2996 non-null   float64
 4   areaWithType     2997 non-null   object 
 5   bedRoom          2997 non-null   object 
 6   bathroom         2997 non-null   object 
 7   balcony          2997 non-null   object 
 8   additionalRoom   1692 non-null   object 
 9   address          2991 non-null   object 
 10  floorNum         2995 non-null   object 
 11  facing           2123 non-null   object 
 12  agePossession    2996 non-null   object 
 13  nearbyLocations  2906 non-null   object 
 14  description      2997 non-null   object 
 15  furnishDetails   2200 non-null   object 
 16  features         2590 non-null   object 
 17  rating           26

#### **4. `bedRoom` Column**

In [55]:
data['bedRoom'].value_counts()

bedRoom
3 Bedrooms    1437
2 Bedrooms     944
4 Bedrooms     478
1 Bedroom      104
5 Bedrooms      31
6 Bedrooms       3
Name: count, dtype: int64

In [60]:
data['bedRoom'] = data['bedRoom'].str.split(' ').str.get(0).str.strip().astype(np.int32)

In [61]:
data['bedRoom'].value_counts()

bedRoom
3    1437
2     944
4     478
1     104
5      31
6       3
Name: count, dtype: int64

In [62]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2997 entries, 0 to 3008
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    2997 non-null   object 
 1   society          2997 non-null   object 
 2   price            2996 non-null   float64
 3   price_per_sqft   2996 non-null   float64
 4   areaWithType     2997 non-null   object 
 5   bedRoom          2997 non-null   int32  
 6   bathroom         2997 non-null   object 
 7   balcony          2997 non-null   object 
 8   additionalRoom   1692 non-null   object 
 9   address          2991 non-null   object 
 10  floorNum         2995 non-null   object 
 11  facing           2123 non-null   object 
 12  agePossession    2996 non-null   object 
 13  nearbyLocations  2906 non-null   object 
 14  description      2997 non-null   object 
 15  furnishDetails   2200 non-null   object 
 16  features         2590 non-null   object 
 17  rating           26

#### **5. `bathroom` Column**

In [63]:
data['bathroom'].value_counts()

bathroom
2 Bathrooms    1044
3 Bathrooms     989
4 Bathrooms     636
5 Bathrooms     169
1 Bathroom      112
6 Bathrooms      42
7 Bathrooms       5
Name: count, dtype: int64

In [69]:
data['bathroom'] = data['bathroom'].str.split(' ').str.get(0).str.strip().astype(np.int32)

#### **6. `balcony` Column**

In [73]:
data['balcony'] = data['balcony'].str.split(' ').str.get(0).str.strip()

In [74]:
data['balcony'].value_counts()

balcony
3     974
3+    862
2     749
1     315
No     97
Name: count, dtype: int64

Replacing **No** to **0**

In [76]:
data['balcony'] = data['balcony'].str.replace('No','0')

In [77]:
data['balcony'].value_counts()

balcony
3     974
3+    862
2     749
1     315
0      97
Name: count, dtype: int64

In [78]:
data.head()

property_name                 society  price  \
0     2 BHK Flat in Krishna Colony  maa bhagwati residency   0.45   
1        2 BHK Flat in Ashok Vihar            apna enclave   0.50   
2              2 BHK Flat in Sohna  tulsiani easy in homes   0.40   
3  2 BHK Flat in Sector 61 Gurgaon     smart world orchard   1.47   
4  2 BHK Flat in Sector 92 Gurgaon        parkwood westend   0.70   

   price_per_sqft                            areaWithType  bedRoom  bathroom  \
0          5000.0          Carpet area: 900 (83.61 sq.m.)        2         2   
1          7692.0          Carpet area: 650 (60.39 sq.m.)        2         2   
2          6722.0          Carpet area: 595 (55.28 sq.m.)        2         2   
3         12250.0        Carpet area: 1200 (111.48 sq.m.)        2         2   
4          5204.0  Super Built up area 1345(124.95 sq.m.)        2         2   

  balcony additionalRoom                              address  \
0       1            NaN     Krishna Colony, Gurgaon, Haryana   
1       1            NaN   46b, Ashok Vihar, Gurgaon, Haryana   
2       3            NaN              Sohna, Gurgaon, Haryana   
3       2     Study Room  Sector 61 Gurgaon, Gurgaon, Haryana   
4       3     Study Room  Sector 92 Gurgaon, Gurgaon, Haryana   

              floorNum facing       agePossession  \
0    4th   of 4 Floors   West     1 to 5 Year Old   
1    1st   of 3 Floors   West        10+ Year Old   
2  12nd   of 14 Floors    NaN     0 to 1 Year Old   
3    2nd   of 4 Floors    NaN              Dec-23   
4    5th   of 8 Floors    NaN  Under Construction   

                                     nearbyLocations  \
0  ['Chintapurni Mandir', 'State bank ATM', 'Pear...   
1  ['Chintapurni Mandir', 'Sheetla Mata Mandir', ...   
2  ['Huda City Metro', 'Golf Course extn road', '...   
3  ['Sector 55-56 Metro station', 'Bestech Centra...   
4  ['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...   

                                         description  \
0  So with lift.Maa bhagwati residency is one of ...   
1  Property situated on main road, railway statio...   
2  This property is 15 km away from badshapur, gu...   
3  Near to metro station of sector 56 and opposit...   
4  We are the proud owners of this 2 bhk alongwit...   

                                      furnishDetails  \
0  ['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...   
1  ['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...   
2                                                NaN   
3                                                NaN   
4                                                 []   

                                            features  \
0  ['Feng Shui / Vaastu Compliant', 'Security / F...   
1  ['Security / Fire Alarm', 'Maintenance Staff',...   
2  ['Power Back-up', 'Feng Shui / Vaastu Complian...   
3  ['Security / Fire Alarm', 'Private Garden / Te...   
4                                                NaN   

                                              rating  
0  ['Environment4 out of 5', 'Safety4 out of 5', ...  
1  ['Environment4 out of 5', 'Safety4 out of 5', ...  
2  ['Environment4 out of 5', 'Safety4 out of 5', ...  
3                                                NaN  
4  ['Environment5 out of 5', 'Safety3 out of 5', ...

#### **7. `additionalRoom` Column**

In [79]:
data['additionalRoom'].value_counts()

additionalRoom
Servant Room                                     629
Study Room                                       232
Others                                           179
Pooja Room                                       132
Study Room,Servant Room                           81
Store Room                                        76
Pooja Room,Servant Room                           60
Servant Room,Others                               52
Servant Room,Pooja Room                           30
Study Room,Others                                 27
Pooja Room,Study Room,Servant Room,Others         25
Pooja Room,Study Room,Servant Room                24
Servant Room,Store Room                           19
Pooja Room,Study Room                             13
Pooja Room,Study Room,Servant Room,Store Room     12
Servant Room,Study Room                            8
Study Room,Pooja Room                              8
Study Room,Servant Room,Store Room                 7
Pooja Room,Servant Room,Others 

In [80]:
# missing values
data['additionalRoom'].isnull().sum()

np.int64(1305)

Replacing the missing values with 'not available' 

In [81]:
data['additionalRoom'].fillna('not available', inplace=True)

In [83]:
data['additionalRoom'] = data['additionalRoom'].str.lower()

In [85]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2997 entries, 0 to 3008
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   property_name    2997 non-null   object 
 1   society          2997 non-null   object 
 2   price            2996 non-null   float64
 3   price_per_sqft   2996 non-null   float64
 4   areaWithType     2997 non-null   object 
 5   bedRoom          2997 non-null   int32  
 6   bathroom         2997 non-null   int32  
 7   balcony          2997 non-null   object 
 8   additionalRoom   2997 non-null   object 
 9   address          2991 non-null   object 
 10  floorNum         2995 non-null   object 
 11  facing           2123 non-null   object 
 12  agePossession    2996 non-null   object 
 13  nearbyLocations  2906 non-null   object 
 14  description      2997 non-null   object 
 15  furnishDetails   2200 non-null   object 
 16  features         2590 non-null   object 
 17  rating           26